libbeat/common/kubernetes/util.go

```diff
@@ -1,68 +1,123 @@
-// it parses the config file to get the config required to build a client.
-func GetKubernetesClient(inCluster bool, kubeConfig string) (client *k8s.Client, err error) {
+// it parses the config file to get the config required to build a clientset.
+func GetKubernetesClientset(inCluster bool, kubeConfig string) (*kubernetes.Clientset, error) {
+	var config *rest.Config
+	var err error
 	if inCluster == true {
-		client, err = k8s.NewInClusterClient()
+		config, err = rest.InClusterConfig()
 		if err != nil {
-			return nil, fmt.Errorf("Unable to get in cluster configuration: %v", err)
+			return nil, err
 		}
 	} else {
-		data, err := ioutil.ReadFile(kubeConfig)
+		config, err = clientcmd.BuildConfigFromFlags("", kubeConfig)
 		if err != nil {
-			return nil, fmt.Errorf("read kubeconfig: %v", err)
+			return nil, err
 		}
+	}
+	config.ContentType = "application/vnd.kubernetes.protobuf"
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return clientset, nil
+}
 
-		// Unmarshal YAML into a Kubernetes config object.
-		var config k8s.Config
-		if err = yaml.Unmarshal(data, &config); err != nil {
-			return nil, fmt.Errorf("unmarshal kubeconfig: %v", err)
+// DiscoverKubernetesNode figures out the Kubernetes node to use.
+// If host is provided in the config use it directly.
+// If beat is deployed in k8s cluster, use hostname of pod which is pod name to query pod meta for node name.
+// If beat is deployed outside k8s cluster, use machine-id to match against k8s nodes for node name.
+func DiscoverKubernetesNode(host string, inCluster bool, clientset *kubernetes.Clientset) (node string) {
+	if host != "" {
+		logp.Info("kubernetes: Using node %s provided in the config", host)
+		return host
+	}
+
+	if inCluster {
+		ns := inClusterNamespace()
+		podName, err := os.Hostname()
+		if err != nil {
+			logp.Err("kubernetes: Couldn't get hostname as beat pod name in cluster with error: ", err.Error())
+			return defaultNode
 		}
-		client, err = k8s.NewClient(&config)
+		logp.Info("kubernetes: Using pod name %s and namespace %s to discover kubernetes node", podName, ns)
+		pod, err := clientset.CoreV1().Pods(inClusterNamespace()).Get(podName, v1.GetOptions{})
 		if err != nil {
-			return nil, err
+			logp.Err("kubernetes: Querying for pod failed with error: ", err.Error())
+			return defaultNode
 		}
+		logp.Info("kubernetes: Using node %s discovered by in cluster pod node query", pod.Spec.NodeSelector)
+		return pod.Spec.NodeName
 	}
 
-	return client, nil
+	mid := machineID()
+	if mid == "" {
+		logp.Err("kubernetes: Couldn't collect info from any of the files in /etc/machine-id /var/lib/dbus/machine-id")
+		return defaultNode
+	}
+
+	nodes, err := clientset.CoreV1().Nodes().List(v1.ListOptions{})
+	if err != nil {
+		logp.Err("kubernetes: Querying for nodes failed with error: ", err.Error())
+		return defaultNode
+	}
+	for _, n := range nodes.Items {
+		if n.Status.NodeInfo.MachineID == mid {
```

Comment:
Cool addition, I guess all the fallback mechanisms make node discovery bulletproof :)